### PR TITLE
api: filter client state updates by worker id

### PR DESCRIPTION
**Pull request by wrboyce**
Change-type: patch
Signed-off-by: Will Boyce \<will@balena.io\>

---
##### Contributor checklist

- [ ] Build output has been rebuilt and tested
- [ ] Introduces security considerations
- [ ] Tests are included
- [ ] Documentation is added or changed
- [ ] Affects the development, build or deployment processes of the component

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -46,6 +46,8 @@ const fromLocalHost: express.RequestHandler = (req, res, next) => {
 export const apiFactory = (serviceId: number) => {
 	const api = express.Router();
 
+	const workerMap: { [key: string]: string } = {};
```
Comment (richbayliss) [inline src/api.ts]:
```suggestion
	const workerMap: _.Dictionary<string> = {};
```

Comment (richbayliss) [reply src/api.ts]: You have `lodash` in your deps anyway, so makes sense to use it's `Dictionary<>` type 👍

File: src/api.ts [hunk 2]
```diff
@@ -64,6 +66,7 @@ export const apiFactory = (serviceId: number) => {
 		}
 		metrics.inc(Metrics.OnlineDevices);
 		metrics.inc(Metrics.TotalDevices);
+		workerMap[req.params.common_name] = req.params.worker;
```
Comment (richbayliss) [inline src/api.ts]: Where does `common_name` come from in the `req.params`?

Comment (wrboyce) [reply src/api.ts]: [this](https://github.com/balena-io/open-balena-vpn/blob/master/openvpn/scripts/client-connect.sh#L26-L33) is called by openvpn when a client connects.

File: src/api.ts [hunk 3]
```diff
@@ -109,6 +112,10 @@ export const apiFactory = (serviceId: number) => {
 			return res.sendStatus(400);
 		}
 
+		if (workerMap[req.body.common_name] !== req.params.worker) {
+			return res.sendStatus(400);
```
Comment (pdcastro) [inline src/api.ts]: Ignore this comment if it is not relevant -- I've jumped in this PR without much context / knowledge of how the VPN service works. But it looks like workerMap is a simple object / dictionary local to the server handling the request. If the requests for `api.post('/api/v2/:worker/clients/')` and `api.delete('/api/v2/:worker/clients/')` were served by different machines (load balancing), then the delete request would be answered with 400.

Comment (wrboyce) [reply src/api.ts]: @pdcastro almost, the `worker id` refers to a process on the machine, not a different physical machine.

Comment (pdcastro) [reply src/api.ts]: But if there were multiple physical machines (maybe there aren't), the issue could still happen... Anyway, it sounds like you're well aware of it, so approving the PR. :-)

File: src/api.ts [hunk 4]
```diff
@@ -109,6 +112,10 @@ export const apiFactory = (serviceId: number) => {
 			return res.sendStatus(400);
 		}
 
+		if (workerMap[req.body.common_name] !== req.params.worker) {
```
Comment (Page-) [inline src/api.ts]: Is it worth adding a sentry exception here so we can see if this happens more than expected? (at least in the short-term)

Comment (Page-) [inline src/api.ts]: Should the workerMap entry be cleared after we've triggered the disconnect?

Comment (wrboyce) [reply src/api.ts]: Great idea, done!

Comment (wrboyce) [reply src/api.ts]: Certainly can't hurt, done.

**Pull request merged**